### PR TITLE
APIM Policy fragments

### PR DESCRIPTION
**Pull request by Hardell**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [x] Other, please clarify - Adding a new feature to a branch with a new API version that's not merged in main yet.
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [x] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json
Comment (j5lim) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json:14]: I see. The API path '../Microsoft.ApiManagement/service/{serviceName}/policyFragments/{id}/references' is for listing references resources under the parent policyFragments resource. The type and id in the response should conform to that.

Can you change it to POST api with something like '/Microsoft.ApiManagement/service/{serviceName}/policyFragments/{id}/listReferences'?

Comment (Hardell) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json:14]: Why POST for a call that just returns a collection? We're not posting any payload nor invoking an action, just getting a collection.

`is for listing references resources under the parent policyFragments resource`
I wouldn't really say that there's a parent/child relationship between those resources. A policy fragment resource can be used by other resources. This API lists those other resources that use a policy fragment. But all of them can exist independently.
E.g. you create a policy fragment called fragment1. Then someone updates other resources(e.g. policy1 and policy2) and includes the policy fragment in them. Calling policyFragments/fragment1/references should then list policy1 and policy2, because they use this policy fragment.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json
Comment (Hardell) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json:37]: Changing this would be a massive breaking change. We can consider fixing this in the future.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json
Comment (j5lim) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json:14]: Listing is a valid POST action. (e.g. listKeys).
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md

Your intent is not a parent/child relationship, but GET on this api path indicates listing all child resources of type `Microsoft.ApiManagement/service/policyFragments/references` and the api should return resources of that type not others since it requires `Microsoft.ApiManagement/service/policyFragments/references/read` RBAC permission.

Having a POST action like `/Microsoft.ApiManagement/service/{serviceName}/policyFragments/{id}/listReferences` or
`/Microsoft.ApiManagement/listPolicyReferences` will avoid the issue.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json
Comment (Hardell) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json:259]: We use the 202 + location header pattern everywhere else in our service. Can we keep the same pattern here for consistency or is the 201/200 + ProvisioningState pattern a requirement?

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json
Comment (Hardell) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json:14]: There are some things I might have explained wrongly or we didn't understand each other:
1. The resources are a child resource of the policy fragment. The type of the resources is the `common-types/resource-management/v2/types.json#/definitions/Resource`. So we need to reply with a collection of the common-type Resource resources that reference the policy fragment. See the `./definitions.json#/definitions/ResourceCollection` for what the definition of this response.
2. The common-type Resource contains Id, Name and Type.  We need those fields to represent the actual resources that reference the policy fragment. Check ApiManagementListPolicyFragmentReferences.json to see an example.
3. We also need pagination (which is currently missing in this PR, I need to add it). If this is a POST request, how do we pass the pagination top and skip parameters? I imagine we shouldn't have a POST request with params in the query string.

Given that we need to list child resources of the type `common-types/resource-management/v2/types.json#/definitions/Resource`, do you still recommend that we make this endpoint a POST?

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json
Comment (adrianhall) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json:259]: (PM here) - we are going to keep our API consistent during the preview version to ensure our service is internally consistent.  I will add a new task to our backlog to update our API en-masse with the new async API pattern.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json
Comment (j5lim) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json:14]: `service/policyFragments/referernces` and `service/policies` are all proxy resources, but it does not mean that the list API for one type can return a collection of the other type. It's a RPC/RBAC violation. The API `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ApiManagement/service/{serviceName}/policyFragments/{id}/references` is for listing `service/policyFragments/referernces` resources.

POST request can have a query string and support pagination.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json
Comment (j5lim) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json:259]: It is okay to use 202. I just mentioned the recommended pattern.

Comment (Hardell) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimpolicyfragments.json:259]: Thanks.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json
Comment (Hardell) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/examples/ApiManagementListPolicyFragmentReferences.json:14]: I've updated the endpoint to be POST and have pagination query string parameters.

**Review by j5lim — APPROVED** [review]
Approved from ARM side.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimanagement.json
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimanagement.json:545]: lets rename this to `PolicyIdFragmentParameter`

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimanagement.json
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimanagement.json:546]: rename from `id` to `policyfragmentid`

**Review by solankisamir — APPROVED** [review]

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimanagement.json
Comment (Hardell) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2021-12-01-preview/apimanagement.json:545]: It was PolicyFragmentIdParameter before, but Maxim asked for this to be generalised so that we can reuse it in the future. 

Most endpoints in APIM have the same ID validation and are already using the new regex validation on the backend but not in this file. This basically introduces a generic ID that can be reused across APIM later on in a PR that fixes the mismatch between this file and the actual implementation.

**Comment by jianyexi** [comment]
@Hardell can you fix the lint errors ?

**Comment by Hardell** [comment]
> @Hardell can you fix the lint errors ?

Due to some constraints that @adrianhall brought up, we made the decision that the CreateOrUpdate PUT request will always return 202 instead of sometimes returning 202 and sometimes being synchronous and returning 200/201 immediately. The result of the operation would be available via a GET request at the URL in the location header(same as before). This is basically the recommended approach you linked for POST requests. Unfortunately, lint doesn't seem to allow that for PUTs.

**Comment by Hardell** [comment]
> > @Hardell can you fix the lint errors ?
> 
> Due to some constraints that @adrianhall brought up, we made the decision that the CreateOrUpdate PUT request will always return 202 instead of sometimes returning 202 and sometimes being synchronous and returning 200/201 immediately. The result of the operation would be available via a GET request at the URL in the location header(same as before). This is basically the recommended approach you linked for POST requests. Unfortunately, lint doesn't seem to allow that for PUTs.

@jianyexi Is there to bypass the lint validation and have it the way described above? Why do we need a terminal state for the PUT when the operation result is returned by a URL in the location header of the PUT response?

**Comment by jianyexi** [comment]
> > > @Hardell
> > >             
> > >             Milan Zolota
> > >             FTE can you fix the lint errors ?
> > 
> > 
> > Due to some constraints that @adrianhall
> >             
> >             Adrian Hall
> >             FTE brought up, we made the decision that the CreateOrUpdate PUT request will always return 202 instead of sometimes returning 202 and sometimes being synchronous and returning 200/201 immediately. The result of the operation would be available via a GET request at the URL in the location header(same as before). This is basically the recommended approach you linked for POST requests. Unfortunately, lint doesn't seem to allow that for PUTs.
> 
> @jianyexi Jianye Xi FTE Is there to bypass the lint validation and have it the way described above? Why do we need a terminal state for the PUT when the operation result is returned by a URL in the location header of the PUT response?

@j5lim  since this is required by ARM that put operation should return 200/201 , can you give some suggestion for this case ?

**Comment by Hardell** [comment]
@j5lim Can you please help?

**Comment by jianyexi** [comment]
@ArcturusZhang  @msyyc  could you help with the SDK generation failure ?

**Comment by Hardell** [comment]
@ArcturusZhang @msyyc Please help with the SDK generation failures.

**Comment by Hardell** [comment]
@jianyexi @ArcturusZhang @msyyc Can we please fix the PR check failures and merge this? There are two failures that seem unrelated to this PR and a third check has been running and stuck for 5 days.

**Comment by Hardell** [comment]
The cross-version breaking change validation is comparing one of my files to a file(stable/2021-08-01/apimanagement.json) that doesn't exist in the target branch. It makes no sense to compare my branch with the main branch because I'm not merging to main.

**Comment by Hardell** [comment]
The other 2 validations are still failing with the same errors that seem unrelated to my changes.

**Comment by solankisamir** [comment]
@jianyexi can we please merge this if all requirements are met? This is a common branch where we are collecting all changes for our new api-version launch.  Much appreciated!

**Comment by jianyexi** [comment]
@Hardell  there is cross version breaking change , you need to do breaking change review by following https://github.com/Azure/azure-rest-api-specs/pull/16403#issuecomment-1007094220
https://github.com/Azure/azure-rest-api-specs/pull/16403/checks?check_run_id=4734551194

**Comment by Hardell** [comment]
@jianyexi 
The cross-version breaking change validation is comparing one of my files to a file(stable/2021-08-01/apimanagement.json) that **doesn't exist in the target branch**. It makes no sense to compare my branch with the main branch because I'm not merging to main. Please, either update the target branch so that it's in sync with the main branch and I can resolve this cross-breaking change, or let me merge.

**Pull request merged**
